File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/legacy/core/PermissionManager.kt
package dev.hebei.camerax_android.legacy.core

import dev.hebei.camerax_android.legacy.CameraXRegistrar
import dev.hebei.camerax_android.legacy.Permission
import dev.hebei.camerax_android.legacy.PigeonApiPermissionManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class PermissionManager(registrar: CameraXRegistrar) : PigeonApiPermissionManager(registrar) {
    override fun instance(): dev.hebei.camerax_android.core.PermissionManager {
        return dev.hebei.camerax_android.core.PermissionManager
    }

    override fun checkPermissioin(
        pigeon_instance: dev.hebei.camerax_android.core.PermissionManager, permission: Permission
    ): Boolean {
        return pigeon_instance.checkPermission(permission.obj)
    }

    override fun requestPermissions(
        pigeon_instance: dev.hebei.camerax_android.core.PermissionManager,
        permissions: List<Permission>,
        callback: (Result<Boolean>) -> Unit
    ) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                val permissionObjs = permissions.map { permission -> permission.obj }
                val isGranted = pigeon_instance.requestPermissions(permissionObjs)
                callback(Result.success(isGranted))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }
}

val Permission.obj
    get() = when (this) {
        Permission.ALBUM -> dev.hebei.camerax_android.core.PermissionManager.Permission.ALBUM
        Permission.AUDIO -> dev.hebei.camerax_android.core.PermissionManager.Permission.AUDIO
        Permission.VIDEO -> dev.hebei.camerax_android.core.PermissionManager.Permission.VIDEO
    }

File: camerax_android/lib/src/legacy/common/permission_manager.dart
import 'package:camerax_android/src/legacy/camerax.g.dart' as $native;
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;

final class PermissionManager extends $interface.PermissionManager {
  final $native.PermissionManager obj;

  PermissionManager.$native(this.obj) : super.impl();

  factory PermissionManager() {
    final obj = $native.PermissionManager.instance;
    return PermissionManager.$native(obj);
  }

  @override
  Future<bool> checkPermission($interface.Permission permission) async {
    final isGranted = await obj.checkPermissioin(permission.obj);
    return isGranted;
  }

  @override
  Future<bool> requestPermissions(
      List<$interface.Permission> permissions) async {
    final permissionObjs =
        permissions.map((permission) => permission.obj).toList();
    final isGranted = await obj.requestPermissions(permissionObjs);
    return isGranted;
  }
}

extension on $interface.Permission {
  $native.Permission get obj {
    switch (this) {
      case $interface.Permission.album:
        return $native.Permission.album;
      case $interface.Permission.audio:
        return $native.Permission.audio;
      case $interface.Permission.video:
        return $native.Permission.video;
    }
  }
}
